azurerm/internal/services/network/express_route_port_resource.go

```diff
@@ -0,0 +1,514 @@
+package network
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/suppress"
+
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/network/validate"
+
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/location"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-05-01/network"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+
+	"github.com/hashicorp/go-azure-helpers/response"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/network/parse"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tags"
+	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+var expressRoutePortSchema = &schema.Schema{
+	Type: schema.TypeList,
+	// Service will always create a pair of links automatically. Users can't add or remove link, but only manipulate existing ones.
+	// This is because the link is actually a map to the physical pair of ports on the MS edge device.
+	Optional: true,
+	Computed: true,
+	MinItems: 1,
+	MaxItems: 1,
+	Elem: &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"admin_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"macsec_cipher": {
+				Type:     schema.TypeString,
+				Optional: true,
+
+				// TODO: The following hardcode can be replaced by SDK types once following is merged:
+				// 	https://github.com/Azure/azure-rest-api-specs/pull/12329
+				Default: "GcmAes128",
+				// Default: string(network.GcmAes128),
+
+				// TODO: The following hardcode can be replaced by SDK types once following is merged:
+				// 	https://github.com/Azure/azure-rest-api-specs/pull/12329
+				ValidateFunc: validation.StringInSlice([]string{
+					"GcmAes128",
+					"GcmAes256",
+					// string(network.GcmAes128),
+					// string(network.GcmAes256),
+				}, false),
+			},
+			"macsec_ckn_keyvault_identifier": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"macsec_cak_keyvault_identifier": {
```

Comment:
would it make sense to group these into a macsec block?

Reply:
It groups things better, whilst we have to make the container block as required. Otherwise, we can allow uses to leave these macsec properties empty and use its default values. This is kind of a trade off, I'm not sure which one is better.

Reply:
i'll leave it up to you, so whichever you think is best

Reply:
I feel it is more easy for users to leave all this empty, rather than setting an empty `macsec {}` block in order to use the default 😅

Reply:
👍